ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp

```diff
@@ -0,0 +1,1210 @@
+// Copyright (C) 2018-2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <algorithm>
+#include <ngraph/opsets/opset7.hpp>
+
+#include "ngraph/runtime/reference/broadcast.hpp"
+#include "ngraph/runtime/reference/matmul.hpp"
+#include "ngraph/runtime/reference/multiply.hpp"
+#include "ngraph/runtime/reference/reshape.hpp"
+#include "ngraph/runtime/reference/sum.hpp"
+#include "ngraph/runtime/reference/transpose.hpp"
+#include "ngraph/shape.hpp"
+
+namespace ngraph
+{
+    namespace runtime
+    {
+        namespace reference
+        {
+            namespace einsum_details
+            {
+                std::vector<std::pair<size_t, size_t>> compute_einsum_path(size_t num_inputs)
```

Comment:
This function looks similar to compute_einsum_path() from einsum_decomposition.cpp. Can a single function be used?

Reply:
Absolutely, agree with you. The situation is not trivial here because we have no common utils routine for both transformations and nGraph core. I have already tried this way but failed. The duplication including three function (`compute_einsum_path`, `is_dimension_reduced`, etc) is about 100 lines not too much. So let us keep these duplications.
Regarding the way to include them into nGraph operation API I consider that `compute_einsum_path` is not finalized (since I mark with TODO) in terms of its API and implementation so we are not ready to include this into nGraph op API.